{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "machine_shape": "hm"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Fungal Genome Annotation Using AUGUSTUS"
      ],
      "metadata": {
        "id": "EQCJmQ-R3713"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "AUGUSTUS is a gene prediction tool widely used in fungal genomics. It uses ab initio prediction combined with species-specific training models to identify protein-coding genes in assembled genomes. This guide walks through the essential steps to annotate a fungal genome using AUGUSTUS."
      ],
      "metadata": {
        "id": "VOvWdwf_4tco"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ukJLb6USUUE9",
        "outputId": "620f2dde-16d7-476a-f52a-de879a3a059c"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "bin/micromamba\n"
          ]
        }
      ],
      "source": [
        "%%bash\n",
        "curl -Ls https://micro.mamba.pm/api/micromamba/linux-64/latest | tar -xvj bin/micromamba\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "./bin/micromamba create -y -n augustus_env -c bioconda -c conda-forge augustus\n"
      ],
      "metadata": {
        "id": "vpmJWd9RWiEl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's annotate a fungal genome Fusarium oxysporum Fo47\n",
        "https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_013085055.1/"
      ],
      "metadata": {
        "id": "TcEPhppOZK1S"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/013/085/055/GCF_013085055.1_ASM1308505v1/GCF_013085055.1_ASM1308505v1_genomic.fna.gz"
      ],
      "metadata": {
        "id": "3jTFaCLAZE1M"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!gunzip GCF_013085055.1_ASM1308505v1_genomic.fna.gz\n",
        "!mv GCF_013085055.1_ASM1308505v1_genomic.fna fungus_genome.fna"
      ],
      "metadata": {
        "id": "xnzL-4GjZE1O"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "eval \"$(./bin/micromamba shell hook -s bash)\"\n",
        "micromamba activate augustus_env\n",
        "augustus --species=fusarium \\\n",
        "         --uniqueGeneId=true \\\n",
        "         --noInFrameStop=true \\\n",
        "         --gff3=on \\\n",
        "         --codingseq=on \\\n",
        "         --protein=on \\\n",
        "         --outfile=Foxy.gff \\\n",
        "         GCF_013085055.1_ASM1308505v1_genomic.fna\n"
      ],
      "metadata": {
        "id": "P2ZNXyjoWjlJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Get the genes and protein sequences of the genome"
      ],
      "metadata": {
        "id": "f01zutI1bszb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!getAnnoFasta.pl Foxy.gff"
      ],
      "metadata": {
        "id": "iUL-nkIObwrP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Rename those files because you will compare those with the masked genome later"
      ],
      "metadata": {
        "id": "TUisPGibb1eR"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Masking the Genome for Repeats in Fungal Annotation\n",
        "Genome masking is a critical preprocessing step before running gene prediction tools such as AUGUSTUS, BRAKER, or MAKER. Fungal genomes frequently contain large amounts of repetitive DNA, including transposable elements, tandem repeats, and low-complexity regions. If these repeats are not masked, annotation tools may incorrectly interpret them as protein-coding genes or exons, leading to false predictions.\n",
        "\n",
        "Why Repeat Masking Matters\n",
        "1. Prevents False Gene Predictions\n",
        "\n",
        "Repetitive sequences often resemble coding regions or exon boundaries.\n",
        "Without masking, gene predictors can:\n",
        "\n",
        "Call false CDS inside transposable elements\n",
        "\n",
        "Introduce spurious introns/exons\n",
        "\n",
        "Inflate gene counts artificially\n",
        "\n",
        "2. Improves Accuracy of Training and Annotation\n",
        "\n",
        "Tools like AUGUSTUS and SNAP use statistical models trained on real genes. Repeats introduce noise that confuses these models.\n",
        "\n",
        "Masking helps ensure:\n",
        "\n",
        "Better exonâ€“intron boundary detection\n",
        "\n",
        "Cleaner gene structures\n",
        "\n",
        "More accurate BUSCO scores\n",
        "\n",
        "3. Essential for High-Quality Comparative Genomics\n",
        "\n",
        "Repeats can inflate:\n",
        "\n",
        "Orthogroup counts\n",
        "\n",
        "Gene family expansions\n",
        "\n",
        "Genome similarity estimates\n",
        "\n",
        "Masked genomes provide more reliable comparative analyses.\n",
        "\n",
        "4. Reduces Computational Load\n",
        "\n",
        "Large repeat regions slow down alignment, evidence integration, and gene calling. Masking simplifies the genome and accelerates annotation."
      ],
      "metadata": {
        "id": "hAspEwNc5rvd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#rename the fast file to maske it\n",
        "!mv GCF_013085055.1_ASM1308505v1_genomic.fna fungus_genome.fna"
      ],
      "metadata": {
        "id": "e_eAEJnv77U-"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Install repeatModeler and repeatMasker"
      ],
      "metadata": {
        "id": "aLH938cRadvD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget -qO- https://micro.mamba.pm/api/micromamba/linux-64/latest \\\n",
        "    | tar -xvj bin/micromamba\n",
        "\n",
        "!mv bin/micromamba /usr/local/bin/micromamba\n",
        "!rmdir bin\n"
      ],
      "metadata": {
        "id": "7nI-MZhU_iEu"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!eval \"$(micromamba shell hook -s bash)\"\n"
      ],
      "metadata": {
        "id": "1pj68L2P_nso"
      },
      "execution_count": 30,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!micromamba create -y -n repeats -c conda-forge -c bioconda \\\n",
        "    repeatmasker repeatmodeler rmblast hmmer trf perl\n",
        "\n"
      ],
      "metadata": {
        "id": "AxkWvToW_q9Q"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!micromamba run -n repeats RepeatMasker -h\n",
        "!micromamba run -n repeats RepeatModeler -h\n",
        "\n"
      ],
      "metadata": {
        "id": "3_sJo9Ru_uvR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!micromamba run -n repeats BuildDatabase -name fungus_db fungus_genome.fna"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "o9qUV0soAX_Z",
        "outputId": "dae47312-9872-4cd5-c98e-297d19b9e505"
      },
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Building database fungus_db:\n",
            "  Reading fungus_genome.fna...\n",
            "Number of sequences (bp) added to database: 12 ( 50358849 bp )\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!micromamba run -n repeats RepeatModeler -database fungus_db -threads 8\n"
      ],
      "metadata": {
        "id": "5f0knuZaAeva"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!ls */consensi.fa.classified\n",
        "!cp RM_9902.MonDec11742392025/consensi.fa.classified ./"
      ],
      "metadata": {
        "id": "Ut9PLUoCWk1t"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!micromamba run -n repeats RepeatMasker \\\n",
        "  -pa 8 \\\n",
        "  -lib consensi.fa.classified \\\n",
        "  -gff \\\n",
        "  fungus_genome.fna\n"
      ],
      "metadata": {
        "id": "3dtFN_voAzIm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!ls -lh fungus_genome.fna*"
      ],
      "metadata": {
        "id": "5bjZx-sqBBzS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!mv */fungus_genome.fna.masked ./"
      ],
      "metadata": {
        "id": "QKqOGyYDXV6s"
      },
      "execution_count": 43,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "fEYEXmsAYXVB"
      },
      "outputs": [],
      "source": [
        "%%bash\n",
        "curl -Ls https://micro.mamba.pm/api/micromamba/linux-64/latest | tar -xvj bin/micromamba\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "./bin/micromamba create -y -n augustus_env -c bioconda -c conda-forge augustus\n"
      ],
      "metadata": {
        "id": "qL3rVpOqYXVC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Annotate the Masked Genome"
      ],
      "metadata": {
        "id": "46S8ZqkFa0BH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "eval \"$(./bin/micromamba shell hook -s bash)\"\n",
        "micromamba activate augustus_env\n",
        "augustus --species=fusarium \\\n",
        "         --uniqueGeneId=true \\\n",
        "         --noInFrameStop=true \\\n",
        "         --gff3=on \\\n",
        "         --codingseq=on \\\n",
        "         --protein=on \\\n",
        "         --outfile=Foxy_unmasked.gff \\\n",
        "        fungus_genome.fna.masked"
      ],
      "metadata": {
        "id": "Lnhkw0sr2LQ7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Get the genes and protein sequences of the masked genome"
      ],
      "metadata": {
        "id": "CtT_vwg-bg2B"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!getAnnoFasta.pl Foxy_unmasked.gff"
      ],
      "metadata": {
        "id": "gY-IzK1tbYVu"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}