{
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# fungal_Effectors\n",
        "Predicting fungal effectors is a key step in understanding host–pathogen interactions, virulence mechanisms, and the molecular strategies fungi use to colonize plant tissues. EffectorP is one of the most widely used machine-learning tools for the identification of fungal effector proteins. It analyzes secreted protein sequences and classifies them based on learned features characteristic of known fungal effectors."
      ],
      "metadata": {
        "id": "4kcCtb8GzVjf"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uELT_rqGrxJn"
      },
      "source": [
        "# Tool Installation\n",
        "First change run type - Go to runtime option and change time option to GPU\n",
        "We'll install the necessary tools, including the source code for EffectorP and Weka, which EffectorP3 relies on as a machine learning dependency. You can access these at:\n",
        "\n",
        "- [EffectorP Source](https://effectorp.csiro.au/)\n",
        "- [Weka (Waikato Environment for Knowledge Analysis)](https://ml.cms.waikato.ac.nz/weka/), an open-source machine learning suite developed by the University of Waikato in New Zealand."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uQtQy0xPrSK-"
      },
      "source": [
        "# Detection of Fungal Effectors\n",
        "\n",
        "This notebook guides you through using EffectorP3, a machine learning tool specifically designed to identify effectors in fungi and oomycetes. Prior to running EffectorP3, it’s essential to pre-filter the input protein sequences, focusing on those likely to be extracellular. Ensuring high confidence in secreted protein predictions enhances the accuracy and relevance of EffectorP3’s results."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "-dlXfxk-bdVy"
      },
      "outputs": [],
      "source": [
        "# Use wget to download the ZIP archive of the repository\n",
        "!wget https://github.com/JanaSperschneider/EffectorP-3.0/archive/refs/heads/main.zip\n",
        "\n",
        "# Unzip the downloaded file\n",
        "!unzip main.zip\n",
        "\n",
        "# Optionally, rename the folder for convenience\n",
        "!mv EffectorP-3.0-main EffectorP-3.0\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "LzgYTODQcVOW"
      },
      "outputs": [],
      "source": [
        "!unzip EffectorP-3.0/weka-3-8-4.zip\n",
        "!mv weka-3-8-4 /content/EffectorP-3.0/"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Install Conda and Seqtk"
      ],
      "metadata": {
        "id": "eUWHIf3HT9RE"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "htJPrSvypuQ0"
      },
      "outputs": [],
      "source": [
        "! wget https://repo.anaconda.com/miniconda/Miniconda3-py37_4.8.2-Linux-x86_64.sh\n",
        "! chmod +x Miniconda3-py37_4.8.2-Linux-x86_64.sh\n",
        "! bash ./Miniconda3-py37_4.8.2-Linux-x86_64.sh -b -f -p /usr/local\n",
        "import sys\n",
        "sys.path.append('/usr/local/lib/python3.7/site-packages/')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7k3IygDOj9gv"
      },
      "outputs": [],
      "source": [
        "!conda install bioconda::seqtk -y\n",
        "!conda install -c conda-forge ncbi-datasets-cli -y"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!datasets download genome accession GCF_013085055.1 --include protein,seq-report\n",
        "!unzip ncbi_dataset.zip\n",
        "!mv ncbi_dataset/data/GCF_013085055.1/protein.faa ./"
      ],
      "metadata": {
        "id": "Tmz0-P8kv5jZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f4qZZxprs-jZ"
      },
      "source": [
        "# Effector prediction\n",
        "Run effectorP"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Hq7T4ua-cYE6"
      },
      "outputs": [],
      "source": [
        "!python EffectorP-3.0/EffectorP.py -i extracelullar.fa > effectorP_output.txt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "w3USu6yatDZ_"
      },
      "source": [
        "# Detect extracelullar proteins using Wolfpsort\n",
        "To ensure high-confidence predictions in protein secretion before running EffectorP3,You can detect extracelullar proteins and prepare you input proteins with WolFPPsort. https://wolfpsort.hgc.jp/aboutWoLF_PSORT.html.en\n",
        "\n",
        "Run WolFPPsort"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bXor9ax7pBzm"
      },
      "outputs": [],
      "source": [
        "!wget https://github.com/fmaguire/WoLFPSort/archive/refs/heads/master.zip\n",
        "!unzip master.zip"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "8xZvxN4vpY0c"
      },
      "outputs": [],
      "source": [
        "# Prompt the user to input the file name\n",
        "input_file = input(\"Enter the path to the input FASTA file (e.g., 'proteins.fa'): \")\n",
        "\n",
        "# Run the WoLFPSort command using the user-provided file name\n",
        "!WoLFPSort-master/bin/runWolfPsortSummary fungi < {input_file} > psortb"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Pdsd7xDStM1W"
      },
      "source": [
        "Prepare your new input and run EffectorP"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "SZRuyhczruuE"
      },
      "outputs": [],
      "source": [
        "#You can detect extracelullar proteins and prepare you input proteins with WolFPPsort. https://wolfpsort.hgc.jp/aboutWoLF_PSORT.html.en\n",
        "\n",
        "input_file = input(\"Enter the path to the input FASTA file (e.g., 'proteins.fa'): \")\n",
        "!cut -f 1,2 -d \" \" psortb | grep extr > extracelullar_list\n",
        "!seqtk subseq {input_file} extracelullar_list > extracelullar.fa"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Run effectorP"
      ],
      "metadata": {
        "id": "Pz9UojnQ0NdT"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dV1jfbQJrGAc"
      },
      "outputs": [],
      "source": [
        "!python EffectorP-3.0/EffectorP.py -i extracelullar.fa > effectorP_output.txt"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TTVWKWo-tRP0"
      },
      "source": [
        "You can use SignalP and TMHMM instead of  WOLPSORT to dtecet extrcaelular proteins"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "FO9bEpkTr9Te"
      },
      "source": [
        "Additional Tools for Preparing Input Data\n",
        "\n",
        "For both SIgnalP and TMHMM, registration and an academic agrement are required. Please request access via the following links:\n",
        "\n",
        "- [SignalP](https://services.healthtech.dtu.dk/cgi-bin/sw_request?software=signalp&version=6.0&packageversion=6.0h&platform=fast)\n",
        "- [TMHMM](https://services.healthtech.dtu.dk/cgi-bin/sw_request?software=tmhmm&version=2.0c&packageversion=2.0c&platform=Linux)\n",
        "\n",
        "After receiving the source code, move it to your workspace and follow the installation instructions provided below. We’ll also use PSortB and WolfPsort as alternatives for localization predictions and Seqtk for efficient sequence parsing."
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "machine_shape": "hm"
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "nbformat": 4,
  "nbformat_minor": 0
}